src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/NamingStyleDiagnosticAnalyzerBase.cs

```diff
@@ -0,0 +1,125 @@
+                SymbolKind.Field,
+                SymbolKind.Method,
+                SymbolKind.NamedType,
+                SymbolKind.Namespace,
+                SymbolKind.Property
+            }.ToImmutableArray();
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(s_descriptorNamingStyle);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(CompilationStartAction);
+        }
+
+        private void CompilationStartAction(CompilationStartAnalysisContext context)
+        {
+            var optionSet = (context.Options as WorkspaceAnalyzerOptions)?.Workspace.Options;
+            var currentValue = optionSet.GetOption(SimplificationOptions.NamingPreferences, context.Compilation.Language);
+
+            if (!string.IsNullOrEmpty(currentValue))
+            {
+                // Deserializing the naming preference info on every CompilationStart is expensive.
+                // Instead, the diagnostic engine should listen for option changes and have the
+                // ability to create the new SerializableNamingStylePreferencesInfo when it detects
+                // any change. The overall system would then only deserialize & allocate when 
+                // actually necessary.
+                var viewModel = Deserialize(currentValue, typeof(SerializableNamingStylePreferencesInfo)) as SerializableNamingStylePreferencesInfo;
+                var preferencesInfo = viewModel.GetPreferencesInfo();
+                context.RegisterSymbolAction(
+                    symbolContext => SymbolAction(symbolContext, preferencesInfo),
+                    _symbolKinds);
+            }
+        }
+
+        private void SymbolAction(SymbolAnalysisContext context, NamingStylePreferencesInfo preferences)
+        {
+            NamingRule applicableRule;
+            if (preferences.TryGetApplicableRule(context.Symbol, out applicableRule))
+            {
+                string failureReason;
+                if (applicableRule.EnforcementLevel != DiagnosticSeverity.Hidden && 
+                    !applicableRule.IsNameCompliant(context.Symbol.Name, out failureReason))
+                {
+                    var descriptor = new DiagnosticDescriptor(IDEDiagnosticIds.NamingRuleId,
+                         "Title", // Todo: strings
+                         string.Format("'{0}' naming violation - ", applicableRule.Title) + failureReason,
+                         DiagnosticCategory.Style,
+                         applicableRule.EnforcementLevel,
+                         isEnabledByDefault: true);
+
+                    var builder = ImmutableDictionary.CreateBuilder<string, string>();
+                    builder[nameof(NamingStyle)] = Serialize(applicableRule.NamingStyle);
+                    builder["OptionName"] = nameof(SimplificationOptions.NamingPreferences); // TODO: doc this
```

Comment:
The core goal here is for the preview pane for a fix to show a link to the relevant option page (or, eventually, to even inline some option-changing UI in the preview pane). It seems like the information about the option that caused the diagnostic should travel with the diagnostic, so that *any* fixes for that diagnostic know about the option (rather than just having *our* fixers know about it), and the information stored needs to be general enough that any host can take the option information and devise its own way to send the user to an option editing view (in VS we have to convert from the option name to the option page Guid, but other hosts may behave differently). Should this be a more proper piece of information on Diagnostics rather than a convention-based member of the property bag?